### PR TITLE
Lazily allocate observables - take 2

### LINKED ISSUE #2387: Lazily allocate observables
E.g. https://github.com/ampproject/amphtml/blob/bf39315ff22225c3e87c8224bfbfb260e08071e4/src/input.js#L75

We should optimize memory usage for cases where never anything registers, or were registration is rare or non-immediate.

### PR DESCRIPTION
**Pull request by chenshay**
close #2387

### CODE REVIEW

File: src/focus-history.js [hunk 1]
```diff
@@ -72,6 +72,9 @@ export class FocusHistory {
    * @return {!UnlistenDef}
    */
   onFocus(handler) {
+    if (!this.observeFocus_) {
```
Comment (cramforce) [inline src/focus-history.js]: Will this always be called on page load?

File: src/input.js [hunk 2]
```diff
@@ -135,6 +138,9 @@ export class Input {
     if (opt_fireImmediately) {
       handler(this.isMouseDetected());
     }
+    if (!this.mouseDetectedObservable_) {
```
Comment (cramforce) [inline src/input.js]: Do we always call these `on` methods on page load?

File: src/service/document-state.js [hunk 3]
```diff
@@ -109,12 +109,17 @@ export class DocumentState {
    * @return {!UnlistenDef}
    */
   onVisibilityChanged(handler) {
+    if (!this.visibilityObservable_) {
```
Comment (cramforce) [inline src/service/document-state.js]: I think this is always called, so should not be lazy.

**Comment by chenshay** [comment]
@cramforce - makes sense. Closing PR

**Pull request closed without merging**